File: Python/clbg/regex-redux/regexredux.py
import re
import sys


def main():
    _ = int(sys.argv[1])

    data = sys.stdin.buffer.read()
    original_length = len(data)

    sequence = re.sub(br'>[^\r\n]*(?:\r\n|\n|\r|$)|\r\n|\r|\n', b'', data)
    stripped_length = len(sequence)

    patterns = [
        b'agggtaaa|tttaccct',
        b'[cgt]gggtaaa|tttaccc[acg]',
        b'a[act]ggtaaa|tttacc[agt]t',
        b'ag[act]gtaaa|tttac[agt]ct',
        b'agg[act]taaa|ttta[agt]cct',
        b'aggg[acg]aaa|ttt[cgt]ccct',
        b'agggt[cgt]aa|tt[acg]accct',
        b'agggta[cgt]a|t[acg]taccct',
        b'agggtaa[cgt]|[acg]ttaccct',
    ]

    output = []
    for pattern in patterns:
        count = sum(1 for _ in re.finditer(pattern, sequence))
        output.append(f'{pattern.decode()} {count}')

    substitutions = [
        (br'tHa[Nt]', b'<4>'),
        (br'aND|caN|Ha[DS]|WaS', b'<3>'),
        (br'a[NSt]|BY', b'<2>'),
        (br'<[^>]*>', b'|'),
        (br'\|[^|][^|]*\|', b'-'),
    ]

    encoded = sequence
    for pattern, replacement in substitutions:
        encoded = re.sub(pattern, replacement, encoded)

    output.append('')
    output.append(str(original_length))
    output.append(str(stripped_length))
    output.append(str(len(encoded)))

    sys.stdout.write('\n'.join(output) + '\n')


if __name__ == '__main__':
    main()

File: Java/clbg/regex-redux/regexredux.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class regexredux {
    private static final String[] VARIANTS = {
        "agggtaaa|tttaccct",
        "[cgt]gggtaaa|tttaccc[acg]",
        "a[act]ggtaaa|tttacc[agt]t",
        "ag[act]gtaaa|tttac[agt]ct",
        "agg[act]taaa|ttta[agt]cct",
        "aggg[acg]aaa|ttt[cgt]ccct",
        "agggt[cgt]aa|tt[acg]accct",
        "agggta[cgt]a|t[acg]taccct"
    };

    private static final String[] SUBSTITUTIONS = {
        "tHa[Nt]",
        "aND|caN|Ha[DS]|WaS",
        "a[NSt]|BY",
        "<[^>]*>",
        "\\|[^|][^|]*\\|"
    };

    private static final String[] REPLACEMENTS = {
        "<4>", "<3>", "<2>", "|", "-"
    };

    private static String readInput() throws IOException {
        return new String(System.in.readAllBytes(), StandardCharsets.ISO_8859_1);
    }

    public static void main(String[] args) throws Exception {
        int n = Integer.parseInt(args[0]);

        String input = readInput();
        int originalLength = input.length();

        final String sequence = Pattern.compile(">[^\\n]*\\n|\\n")
                .matcher(input)
                .replaceAll("");
        int strippedLength = sequence.length();
        input = null;

        int workers = Math.max(1, Math.min(
                VARIANTS.length, Runtime.getRuntime().availableProcessors()));
        ExecutorService executor = Executors.newFixedThreadPool(workers);

        try {
            List<Future<Integer>> counts = new ArrayList<>(VARIANTS.length);

            for (String variant : VARIANTS) {
                counts.add(executor.submit(() -> {
                    Matcher matcher = Pattern.compile(variant).matcher(sequence);
                    int count = 0;
                    while (matcher.find()) {
                        count++;
                    }
                    return count;
                }));
            }

            String encoded = sequence;
            for (int i = 0; i < SUBSTITUTIONS.length; i++) {
                encoded = Pattern.compile(SUBSTITUTIONS[i])
                        .matcher(encoded)
                        .replaceAll(REPLACEMENTS[i]);
            }

            StringBuilder output = new StringBuilder(512);
            for (int i = 0; i < VARIANTS.length; i++) {
                output.append(VARIANTS[i]).append(' ')
                        .append(counts.get(i).get()).append('\n');
            }

            output.append(originalLength).append('\n')
                    .append(strippedLength).append('\n')
                    .append(encoded.length()).append('\n');

            System.out.print(output);
        } finally {
            executor.shutdown();
        }
    }
}